File: src/commonutils/bigdata/HDFSUtil.scala
package com.haiyisoft.utils

import java.net.URI

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.permission.{FsAction, FsPermission}
import org.apache.hadoop.fs.{FileStatus, FileSystem, FileUtil, Path}

import scala.collection.mutable.ArrayBuffer

class HDFSUtil(hdfsBasePath: String) {

  System.setProperty("HADOOP_USER_NAME", "hdfs")
  System.setProperty("user.name", "hdfs")
  private val fsUri = new URI(hdfsBasePath)
  private val fs = FileSystem.get(fsUri, new Configuration())

  //设置HDFS用户
  def setUser(user: String): Unit = {
    System.setProperty("HADOOP_USER_NAME", user)
    System.setProperty("user.name", user)
  }

  /**
    * 列出目录下所有文件,不包括目录
    *
    * @param path
    * @return
    */
  def listFiles(path: String): Array[String] = {
    fs.listStatus(getPath(path)).map(_.getPath.toUri.getPath)
  }

  /**
    * 列出目录下所有目录
    *
    * @param path
    * @param recursive
    * @return
    */
  def listDirs(path: String, recursive: Boolean = false): Array[String] = {
    val arr = ArrayBuffer[String]()
    val paths = fs.listStatus(getPath(path)).filter(_.isDirectory).map(_.getPath.toUri.getPath)
    if (recursive) {
      paths.foreach { x =>
        listDirs(x, recursive).foreach(arr.append(_))
        arr.append(x)
      }
    } else {
      paths.foreach(arr.append(_))
    }
    arr.toArray
  }

  /**
    * 列出最底层目录
    *
    * @param path
    * @return
    */
  def listBottomDirs(path: String): Array[String] = {
    val dirs = listDirs(path, true)
    dirs.filter { x =>
      listDirs(x, true).length == 0
    }
  }

  /**
    * 列出所有目录
    *
    * @param path 目录
    * @param recursive 递归
    * @return
    */
  def listAllFiles(path: String, recursive: Boolean): Array[String] = {
    val arr = ArrayBuffer[String]()
    val files = fs.listFiles(getPath(path), recursive)
    while (files.hasNext) {
      val f = files.next()
      if (f.isFile)
        arr.append(f.getPath.toUri.getPath)
    }
    arr.toArray
  }

  /**
    * 删除目录
    *
    * @param paths
    */
  def deleteDir(paths: String*): Unit = {
    paths.par.foreach { path =>
      if (fs.exists(getPath(path)) && fs.isDirectory(getPath(path))) {
        println(s"try to delete $path")
        val dirs = listDirs(path)
        deleteDir(dirs: _*)
        val files = listAllFiles(path, recursive = false)
        deleteFile(files: _*)
        fs.delete(getPath(path), true)
      } else {
        println(s"$path is not a Dir or not exits")
      }
    }
  }

  /**
    * 删除文件
    *
    * @param paths
    */
  def deleteFile(paths: String*): Unit = {
    paths.par.foreach { path =>
      if (fs.exists(getPath(path)) && !fs.isDirectory(getPath(path))) {
        println(s"try to delete $path")
        val res = fs.delete(getPath(path), true)
        if (res) {
          println(s"complete deleting $path")
        } else {
          println(s"an error occurred during deleting $path")
        }
      }
      else {
        println(s"$path is not a Dir")
      }
    }
  }

  /**
    * 移动
    *
    * @param source
    * @param dest
    */
  def move(source: String, dest: String): Unit = {
    copy(source, dest, deleteSourceFile = true)
  }

  /**
    * 复制
    *
    * @param source 源地址
    * @param destination 目标地址
    * @param deleteSourceFile 是否删除源文件
    * @param overWrite 是否覆盖
    * @return
    */
  def copy(source: String, destination: String, deleteSourceFile: Boolean = false, overWrite: Boolean = false): Boolean = {
    val sourceFile = getPath(source)
    val destFile = getPath(destination)

    if (!fs.exists(destFile)) {
      fs.mkdirs(destFile)
    }

    if (fs.exists(sourceFile) && fs.exists(destFile)) {
      if (fs.isDirectory(sourceFile) && fs.isDirectory(destFile)) {
        val sourceDir = new FileStatus()
        sourceDir.setPath(sourceFile)
        FileUtil.copy(fs, sourceDir, fs, destFile, deleteSourceFile, overWrite, new Configuration())
      } else if (fs.isFile(sourceFile) && fs.isDirectory(destFile)) {
        FileUtil.copy(fs, sourceFile, fs, destFile, deleteSourceFile, overWrite, new Configuration())
      } else if (fs.isFile(sourceFile) && fs.isFile(destFile)) {
        FileUtil.copy(fs, sourceFile, fs, destFile, deleteSourceFile, overWrite, new Configuration())
      } else {
        println("cant copy a directory to a file")
        false
      }
    } else {
      println("sorce file or destination file not exits")
      false
    }
  }

  /**
    * 获取HDFS地址
    *
    * @param path
    * @return
    */
  def getPath(path: String): Path = {
    //    if (path.toLowerCase().startsWith("hdfs://")) {
    //      new Path(path)
    //    } else {
    //      new Path(hdfsBasePath + path)
    //    }
    new Path(path)
  }

  /**
    * 改变所属用户
    *
    * @param path      路径
    * @param group     所属组
    * @param owner     所属用户
    * @param recursive 是否递归
    */
  def chown(path: String, group: String, owner: String, recursive: Boolean = true): Unit = {
    val p = getPath(path)
    val oldGroup = fs.getFileStatus(p).getGroup
    val oldOwner = fs.getFileStatus(p).getOwner

    if (group.equals(oldGroup) && owner.equals(oldOwner)) {
      println("组名和用户名未改变")
    } else {
      try {
        fs.setOwner(p, owner, group)
        if (recursive && fs.getFileStatus(p).isDirectory) {
          listDirs(path).par.foreach(p1 => chown(p1, group, owner, recursive))
          listAllFiles(path, recursive = false).par.foreach(p1 => chown(p1, group, owner, recursive = false))
        }
        println(s"changed owner for [$path]")
      } catch {
        case e: Exception =>
          e.printStackTrace()
      }
    }
  }

  /**
    * 改变权限
    *
    * 0 无
    * 1 可执行
    * 2 写
    * 4 读
    * 3 可执行 + 写
    * 5 可执行 + 读
    * 6 读 + 写
    * 7 读 + 写 + 可执行
    *
    * @param path            地址
    * @param userPermission  所属用户权限
    * @param groupPermission 所属组权限
    * @param otherPermission 其他用户权限
    * @param recursive       是否递归
    */
  def chmod(path: String, userPermission: Int, groupPermission: Int, otherPermission: Int, recursive: Boolean): Unit = {

    require(userPermission <= 7 && userPermission > 0
      && groupPermission <= 7 && groupPermission > 0
      && otherPermission <= 7 && otherPermission > 0, "Invalid permission!")

    /**
      * 根据数字获取权限信息
      *
      * @param p
      * @return
      */
    def getPermission(p: Int): FsAction = {
      if (p == 0) {
        FsAction.NONE
      } else if (p == 1) {
        FsAction.EXECUTE
      } else if (p == 2) {
        FsAction.WRITE
      } else if (p == 3) {
        FsAction.WRITE_EXECUTE
      } else if (p == 4) {
        FsAction.READ
      } else if (p == 5) {
        FsAction.READ_EXECUTE
      } else if (p == 6) {
        FsAction.READ_WRITE
      } else {
        FsAction.ALL
      }
    }

    if (fs.isDirectory(getPath(path)) && recursive) {
      val dirs = listDirs(path)
      if (dirs.length > 0) {
        dirs.par.foreach { p =>
          chmod(p, userPermission, groupPermission, otherPermission, recursive)
        }
      }
      val files = listAllFiles(path, recursive = false)
      if (files.length > 0) {
        files.par.foreach { p =>
          chmod(p, userPermission, groupPermission, otherPermission, recursive)
        }
      }
    }

    val m: FsPermission = new FsPermission(getPermission(userPermission), getPermission(groupPermission), getPermission(otherPermission))

    fs.setPermission(getPath(path), m)
  }

  /**
    * 改变权限
    *
    * @param path 路径
    * @param mod 权限 形如 "777", "766", "664" 的形式
    * @param recursive 递归
    */
  def chmod(path: String, mod: String, recursive: Boolean = true): Unit = {
    require(mod.length > 0 && mod.length <= 3, s"Invalid permission number, required 3 but found ${mod.length}!")
    mod.foreach { x =>
      val c = x.toInt
      require(c >= 48 && c <= 55, s"Invalid permisson at number $x")
    }
    val arr = mod.split("")
    val u = arr(0).toInt
    val g = arr(1).toInt
    val o = arr(2).toInt
    chmod(path, u, g, o, recursive)
  }

  /**
    * 创建目录
    *
    * @param path
    * @return
    */
  def mkDir(path: String): Boolean = {
    if (fs.exists(getPath(path))) {
      println(s"$path already exits")
      false
    } else {
      fs.mkdirs(getPath(path))
    }
  }

  /**
    * 创建文件
    *
    * @param path
    * @return
    */
  def createFile(path: String): Boolean = {
    if (fs.exists(getPath(path))) {
      println(s"$path already exits")
      false
    } else {
      fs.createNewFile(getPath(path))
    }
  }

  /**
    * 重命名
    *
    * @param path1 原地址
    * @param path2 目标地址
    */
  def rename(path1: String, path2: String): Unit = {
    fs.rename(getPath(path1), getPath(path2))
  }

  /**
    * 地址是否存在
    *
    * @param path
    * @return
    */
  def exists(path: String): Boolean = {
    fs.exists(getPath(path))
  }
}

/**
  * 伴生
  */
object HDFSUtil {
  def getInstance(baseUrl: String): HDFSUtil = {
    new HDFSUtil(baseUrl)
  }
}

File: src/zodiac/java/commonutils/hadoop/HdfsUtil.java
package zodiac.java.commonutils.hadoop;

import org.apache.commons.lang3.StringUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.*;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import zodiac.java.commonutils.text.UnitUtil;

import java.io.*;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class HdfsUtil {
    private String url;
    private String user = "hdfs";
    private FileSystem fs;
    private final Logger logger = LogManager.getLogger(this.getClass());
    private Charset defaultCharset = StandardCharsets.UTF_8;

    public HdfsUtil(String url) {
        assert StringUtils.isNotBlank(url);
        this.url = url;
        init();
    }

    public HdfsUtil(String url, String user) {
        assert StringUtils.isNotBlank(url) && StringUtils.isNotBlank(user);
        this.url = url;
        this.user = user;
        init();
    }

    private void init() {
        assert url.startsWith("hdfs://") || url.startsWith("/");
        Pattern pattern = Pattern.compile("^(hdfs)://(([\\d]{1,3}.[\\d]{1,3}.[\\d]{1,3}.[\\d]{1,3})|([1-9a-zA-Z.]+)):([\\d]{1,5})");
        Matcher matcher = pattern.matcher(url);
        if (matcher.matches()) {
            url = matcher.group(0);
        } else {
            throw new IllegalArgumentException("Base hdfs url is invalid!");
        }
        System.setProperty("HADOOP_USER_NAME", user);
        System.setProperty("user.name", user);
        try {
            this.fs = FileSystem.get(new URI(url), new Configuration(true));
        } catch (IOException e) {
            logger.error("Error when initialize hdfs file system!", e);
            e.printStackTrace();
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
    }

    public void setUser(String user) {
        this.user = user;
    }

    public void setDefaultCharset(Charset defaultCharset) {
        this.defaultCharset = defaultCharset;
    }

    public String[] list(String path) {
        return list(path, (Objects::nonNull));
    }

    public String[] listFiles(String path) {
        return list(path, FileStatus::isFile);
    }

    public String[] listDirs(String path) {
        return list(path, FileStatus::isDirectory);
    }

    public String[] list(String path, Predicate<FileStatus> fileFilter) {
        ArrayList<String> ret = new ArrayList<>();
        if (exists(path)) {
            if (isDirectory(path)) {
                try {
                    ret.addAll(Arrays.stream(fs.listStatus(getPath(path))).parallel().filter(fileFilter)
                            .map(x -> x.getPath().toUri().getRawPath()).collect(Collectors.toCollection(ArrayList::new)));
                } catch (IOException e) {
                    logger.error("Error when get file list!", e);
                }
            }
        }
        return ret.toArray(new String[0]);
    }

    public String[] listAll(String path) {
        ArrayList<String> ret = new ArrayList<>();
        for (String s : list(path)) {
            ret.add(s);
            if (isDirectory(s)) {
                ret.addAll(Arrays.asList(listAll(s)));
            }
        }
        return ret.toArray(new String[0]);
    }

    public String[] listAllFiles(String path) {
        ArrayList<String> ret = new ArrayList<>();
        for (String s : list(path)) {
            if (isFile(s)) {
                ret.add(s);
            } else {
                ret.addAll(Arrays.asList(listAllFiles(s)));
            }
        }
        return ret.toArray(new String[0]);
    }

    public String[] listAllDirs(String path) {
        ArrayList<String> ret = new ArrayList<>();
        for (String s : list(path)) {
            if (isDirectory(s)) {
                ret.add(s);
                ret.addAll(Arrays.asList(listAllDirs(s)));
            }
        }
        return ret.toArray(new String[0]);
    }

    public boolean delete(String... path) {
        assert null != path && path.length > 0 : "Paths should not be empty or null!";
        boolean ret = true;
        for (String s : path) {
            try {
                ret &= fs.delete(getPath(s), true);
            } catch (IOException e) {
                logger.error("Error when delete " + s, e);
            }
        }
        return ret;
    }

    public boolean exists(String path) {
        boolean exits = false;
        try {
            exits = fs.exists(getPath(path));
        } catch (IOException e) {
            logger.error(e);
        }
        return exits;
    }

    public boolean createFile(String path) {
        boolean created = false;
        if (!exists(path)) {
            try {
                created = fs.createNewFile(getPath(path));
            } catch (IOException e) {
                logger.error("Error when create file " + path, e);
            }
        }
        return created;
    }

    public boolean mkdir(String path) {
        boolean created = false;
        if (!exists(path)) {
            try {
                created = fs.mkdirs(getPath(path));
            } catch (IOException e) {
                logger.error("Error when create file " + path, e);
            }
        }
        return created;
    }

    public boolean chown(String user, String group, String... path) {
        boolean ch = true;
        for (String s : path) {
            try {
                if (StringUtils.isBlank(group)) {
                    fs.setOwner(getPath(s), user, getFileStatus(s).getGroup());
                } else {
                    fs.setOwner(getPath(s), user, group);
                }
            } catch (Exception e) {
                ch = false;
                logger.error("Error when chown " + s, e);
            }
        }
        return ch;
    }

    public boolean chown(String user, String... path) {
        return chown(user, null, path);
    }

    public FileStatus getFileStatus(String path) {
        FileStatus ret = null;
        try {
            ret = fs.getFileStatus(getPath(path));
        } catch (IOException e) {
            logger.error("Error when get file status: " + path, e);
        }
        return ret;
    }

    public boolean rename(String sourcePath, String destPath) {
        return rename(getPath(sourcePath), getPath(destPath));
    }

    public boolean rename(Path source, Path dest) {
        boolean ret = false;
        try {
            ret = fs.rename(source, dest);
        } catch (IOException e) {
            logger.error("Error when rename " + source + " to " + dest, e);
        }
        return ret;
    }

    public boolean move(Path source, Path dest) {
        return rename(source, dest);
    }

    public boolean copy(String sourcePath, String destPath, boolean overwrite) {
        boolean c = false;
        try {
            c = FileUtil.copy(fs, getPath(sourcePath), fs, getPath(destPath), false, overwrite, new Configuration());
        } catch (IOException e) {
            logger.error("Error when copy " + sourcePath + " to " + destPath, e);
        }
        return c;
    }

    public boolean move(String sourcePath, String destPath) {
        return rename(sourcePath, destPath);
    }

    public boolean moveToTrash(String... paths) {
        boolean ret = true;
        for (String path : paths) {
            Path p = getPath(path);
            Path trashPath = fs.getTrashRoot(p);
            ret = ret && move(p, trashPath);
        }
        return ret;
    }

    public boolean chmod(String path, String mod) {
        boolean b = false;
        try {
            b = FileUtil.chmod(path, mod, true) == 0;
        } catch (IOException e) {
            logger.error("Error when chmod " + path, e);
        }
        return b;
    }

    public boolean put(String sourcePath, String destPath) {
        boolean b = false;
        try {
            fs.copyFromLocalFile(getPath(sourcePath), getPath(destPath));
            b = true;
        } catch (IOException e) {
            logger.error("Error when copy from " + sourcePath + " to " + destPath, e);
        }
        return b;
    }

    public boolean get(String sourcePath, String destPath) {
        boolean b = false;
        try {
            fs.copyToLocalFile(getPath(sourcePath), getPath(destPath));
            b = true;
        } catch (IOException e) {
            logger.error("Error when copy from " + sourcePath + " to " + destPath, e);
        }
        return b;
    }

    public boolean isFile(String path) {
        return getFileStatus(path).isFile();
    }

    public boolean isDirectory(String path) {
        return getFileStatus(path).isDirectory();
    }

    private Path getPath(String path) {
        return new Path(path);
    }

    public String getParent(String path) {
        return getPath(path).getParent().toUri().getRawPath();
    }

    public String getName(String path) {
        return getPath(path).getName();
    }

    public long getSize(String path) {
        return getFileStatus(path).getLen();
    }

    public long getTotalFileSize(String... paths) {
        long total = 0;
        for (String path : paths) {
            if (isFile(path)) {
                total += getSize(path);
            } else {
                for (String s : listAllFiles(path)) {
                    total += getSize(s);
                }
            }
        }
        return total;
    }

    public String getTotalFileSizeAsHumanReadable(String... paths) {
        return UnitUtil.ByteUtil.toString(getTotalFileSize(paths), true);
    }

    public boolean saveObject(Object o, String path, boolean overwrite) {
        boolean res = false;
        try {
            FSDataOutputStream dataOutputStream = fs.create(getPath(path), overwrite);
            ObjectOutputStream objectOutputStream = new ObjectOutputStream(dataOutputStream.getWrappedStream());
            objectOutputStream.writeObject(o);
            objectOutputStream.flush();
            dataOutputStream.close();
            objectOutputStream.close();
            res = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return res;
    }

    public Object readObject(String path) {
        Object res = null;
        try {
            ObjectInputStream objectInputStream = new ObjectInputStream(fs.open(getPath(path)));
            res = objectInputStream.readObject();
            objectInputStream.close();
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
        }
        return res;
    }

    public FSDataInputStream getInputStream(String path) {
        FSDataInputStream inputStream = null;
        try {
            inputStream = fs.open(getPath(path));
        } catch (IOException e) {
            e.printStackTrace();
        }
        return inputStream;
    }

    public byte[] readAsBytes(String path) {
        return readAsBytes(path, Integer.MAX_VALUE);
    }

    private byte[] readAsBytes(String path, int len) {
        byte[] result = null;
        FSDataInputStream inputStream = getInputStream(path);
        if (len < 0) {
            throw new IllegalArgumentException("len < 0");
        }
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        int read;
        try {
            int c = 0;
            while ((read = inputStream.read()) != -1 && c < len) {
                byteArrayOutputStream.write(read);
                c++;
            }
            result = byteArrayOutputStream.toByteArray();
            byteArrayOutputStream.close();
            inputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }

    public ArrayList<String> readAsString(String path, String charset) {
        BufferedReader bufferedReader;
        ArrayList<String> result = new ArrayList<>();
        try {
            bufferedReader = new BufferedReader(new InputStreamReader(getInputStream(path), charset));
            result = bufferedReader.lines().collect(Collectors.toCollection(ArrayList::new));
            bufferedReader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return result;
    }

    public ArrayList<String> readAsString(String path) {
        return readAsString(path, "utf-8");
    }

    public String readAsStringQuickly(String path, String charset) {
        String result = null;
        try {
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            FSDataInputStream fsDataInputStream = getInputStream(path);
            byte[] buffer = new byte[1024];
            int read;
            while ((read = fsDataInputStream.read(buffer)) != -1) {
                byteArrayOutputStream.write(buffer, 0, read);
            }
            result = byteArrayOutputStream.toString(charset);
            byteArrayOutputStream.close();
            fsDataInputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }

    public String readAsStringQuickly(String path) {
        return readAsStringQuickly(path, defaultCharset.name());
    }

    public FSDataOutputStream getOutputStream(String path, boolean overwrite) {
        FSDataOutputStream outputStream = null;
        try {
            outputStream = fs.create(getPath(path), overwrite);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return outputStream;
    }

    public FSDataOutputStream getOutputStream(String path) {
        return getOutputStream(path, true);
    }

    public void writeBytes(String path, byte[] bytes) {
        OutputStream outputStream = getOutputStream(path);
        try {
            outputStream.write(bytes);
            outputStream.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void close() {
        if (null != fs) {
            try {
                fs.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

}
